File: minimoog/minimoog_instrument/minimoog_instrumentAudioUnit.m
//
//  minimoog_instrumentAudioUnit.m
//  minimoog_instrument
//
//  Created by Yauheni Lychkouski on 10/6/18.
//  Copyright © 2018 Yauheni Lychkouski. All rights reserved.
//

#import "minimoog_instrumentAudioUnit.h"

#import <AVFoundation/AVFoundation.h>

// Define parameter addresses.
enum {
    osc1RangeParamId = 0,
    osc1WaveformParamId,
    osc2RangeParamId,
    osc2DetuneParamId,
    osc2WaveformParamId,
    mixOsc1VolumeParamId,
    mixOsc2VolumeParamId,
    mixNoiseVolumeParamId,
    lastParamId = -1
};


typedef struct {
    AudioUnitParameterID paramId;
    const char *identifier;
    const char *name;
    float min;
    float max;
    float initVal;
    AudioUnitParameterUnit unit;
    const char *commaSeparatedIndexedNames;
} ParameterDef;


const ParameterDef paramDef[] = {
    {osc1RangeParamId     , "osc1Range"     , "Oscillator 1 Range"       ,  0,  5,  0, kAudioUnitParameterUnit_Indexed, "LO,32',16',8',4',2'" },
    {osc1WaveformParamId  , "osc1Waveform"  , "Oscillator 1 Waveform"    ,  0,  5,  0, kAudioUnitParameterUnit_Indexed, "Triangle,Ramp,Sawtooth,Square,Pulse1,Pulse2" },
    {osc2RangeParamId     , "osc2Range"     , "Oscillator 2 Range"       ,  0,  5,  0, kAudioUnitParameterUnit_Octaves, "LO,32',16',8',4',2'" },
    {osc2DetuneParamId    , "osc2Detune"    , "Oscillator 2 Detune"      , -8,  8,  0, kAudioUnitParameterUnit_Cents  , "" },
    {osc2WaveformParamId  , "osc2Waveform"  , "Oscillator 2 Waveform"    ,  0,  5,  0, kAudioUnitParameterUnit_Indexed,  "Triangle,Ramp,Sawtooth,Square,Pulse1,Pulse2" },
    {mixOsc1VolumeParamId , "mixOsc1Volume" , "Mixer Oscillator 1 Volume",  0, 10, 10, kAudioUnitParameterUnit_CustomUnit, "" },
    {mixOsc2VolumeParamId , "mixOsc2Volume" , "Mixer Oscillator 2 Volume",  0, 10,  0, kAudioUnitParameterUnit_CustomUnit, "" },
    {mixNoiseVolumeParamId, "mixNoiseVolume", "Mixer Noise Volume"       ,  0, 10,  0, kAudioUnitParameterUnit_CustomUnit, "" },
    {lastParamId}
};


@interface minimoog_instrumentAudioUnit ()
    @property AUAudioUnitBus *outputBus;
    @property AUAudioUnitBusArray *outputBusArray;
    @property (nonatomic, readwrite) AUParameterTree *parameterTree;
@end


@implementation minimoog_instrumentAudioUnit
    @synthesize outputBus;
    @synthesize outputBusArray;
    @synthesize parameterTree;

- (instancetype)initWithComponentDescription:(AudioComponentDescription)componentDescription options:(AudioComponentInstantiationOptions)options error:(NSError **)outError {
    self = [super initWithComponentDescription:componentDescription options:options error:outError];
    
    if (self == nil) {
        return nil;
    }
    
    // Create parameter objects.
    NSMutableArray *params = [NSMutableArray array];
    int i = 0;
    while (paramDef[i].paramId != lastParamId) {
        AUParameter *param =
            [AUParameterTree
             createParameterWithIdentifier:[NSString stringWithUTF8String:paramDef[i].identifier]
             name:[NSString stringWithUTF8String:paramDef[i].name]
             address:paramDef[i].paramId
             min:paramDef[i].min
             max:paramDef[i].max
             unit:paramDef[i].unit
             unitName:nil
             flags:0
             valueStrings:[[NSString stringWithUTF8String:paramDef[i].commaSeparatedIndexedNames] componentsSeparatedByString:@","]
             dependentParameters:nil];
        
        // Initialize the parameter values.
        param.value = paramDef[i].initVal;
        
        [params addObject:param];
        
        i++;
    }
    
    
    
    
    _kernel.setParameter(InstrumentParamAttack, attackParam.value);
    _kernel.setParameter(InstrumentParamRelease, releaseParam.value);
    
    
    
    
    // Create the parameter tree.
    self.parameterTree = [AUParameterTree createTreeWithChildren:params];
    
    
    
    
    // Create the output bus.
    AVAudioFormat *defaultFormat = [[AVAudioFormat alloc] initStandardFormatWithSampleRate:44100. channels:2];
    self.outputBusBuffer.init(defaultFormat, 2);
    self.outputBus = self.outputBusBuffer.bus;
    
    // Create the input and output bus arrays.
    outputBusArray = [[AUAudioUnitBusArray alloc] initWithAudioUnit:self
                                                            busType:AUAudioUnitBusTypeOutput
                                                             busses: @[outputBus]];
    
    // Make a local pointer to the kernel to avoid capturing self.
    __block InstrumentDSPKernel *instrumentKernel = &_kernel;
    
    // implementorValueObserver is called when a parameter changes value.
    _parameterTree.implementorValueObserver = ^(AUParameter *param, AUValue value) {
        instrumentKernel->setParameter(param.address, value);
    };
    
    // implementorValueProvider is called when the value needs to be refreshed.
    _parameterTree.implementorValueProvider = ^(AUParameter *param) {
        return instrumentKernel->getParameter(param.address);
    };
    
    // A function to provide string representations of parameter values.
    _parameterTree.implementorStringFromValueCallback = ^(AUParameter *param, const AUValue *__nullable valuePtr) {
        AUValue value = valuePtr == nil ? param.value : *valuePtr;
        
        switch (param.address) {
            case InstrumentParamAttack:
            case InstrumentParamRelease:
                return [NSString stringWithFormat:@"%.3f", value];
                
            default:
                return @"?";
        }
    };
    
    
    
    
    
    self.maximumFramesToRender = 512;
    
    return self;
}

#pragma mark - AUAudioUnit Overrides

/*
 Overriding AUAudioUnit Properties and Methods
 You must override the following properties in your AUAudioUnit subclass:
 Override the inputBusses getter method to return the app extension’s audio input connection points.
 Override the outputBusses getter method to return the app extension’s audio output connection points.
 Override the internalRenderBlock getter method to return the block that implements the app extension’s audio rendering loop.
 Also override the allocateRenderResourcesAndReturnError: method, which the host app calls before it starts to render audio, and override the deallocateRenderResources method, which the host app calls after it has finished rendering audio. Within each override, call the AUAudioUnit superclass implementation.
 */





// If an audio unit has input, an audio unit's audio input connection points.
// Subclassers must override this property getter and should return the same object every time.
// See sample code.
- (AUAudioUnitBusArray *)inputBusses {
#warning implementation must return non-nil AUAudioUnitBusArray
    return nil;
}

// An audio unit's audio output connection points.
// Subclassers must override this property getter and should return the same object every time.
// See sample code.
- (AUAudioUnitBusArray *)outputBusses {
#warning implementation must return non-nil AUAudioUnitBusArray
    return nil;
}

// Allocate resources required to render.
// Subclassers should call the superclass implementation.
- (BOOL)allocateRenderResourcesAndReturnError:(NSError **)outError {
    if (![super allocateRenderResourcesAndReturnError:outError]) {
        return NO;
    }
    
    // Validate that the bus formats are compatible.
    // Allocate your resources.
    
    return YES;
}

// Deallocate resources allocated in allocateRenderResourcesAndReturnError:
// Subclassers should call the superclass implementation.
- (void)deallocateRenderResources {
    // Deallocate your resources.
    [super deallocateRenderResources];
}

#pragma mark - AUAudioUnit (AUAudioUnitImplementation)

// Block which subclassers must provide to implement rendering.
- (AUInternalRenderBlock)internalRenderBlock {
    // Capture in locals to avoid Obj-C member lookups. If "self" is captured in render, we're doing it wrong. See sample code.
    
    return ^AUAudioUnitStatus(AudioUnitRenderActionFlags *actionFlags, const AudioTimeStamp *timestamp, AVAudioFrameCount frameCount, NSInteger outputBusNumber, AudioBufferList *outputData, const AURenderEvent *realtimeEventListHead, AURenderPullInputBlock pullInputBlock) {
        // Do event handling and signal processing here.
        
        return noErr;
    };
}

@end

File: minimoog/minimoog_instrument/MinimoogInstrumentAudioUnit.swift
//
//  MinimoogInstrumentAudioUnit.m
//
//  Created by Yauheni Lychkouski on 10/6/18.
//  Copyright © 2018 Yauheni Lychkouski. All rights reserved.
//

import AVFoundation

extension String : Error { }

public class MinimoogInstrumentAudioUnit : AUAudioUnit {
    // MARK: Types
    enum ParamAddr : AUParameterAddress, CaseIterable {
        case osc1RangeParamAddr = 0
        case osc1WaveformParamAddr
        case osc2RangeParamAddr
        case osc2DetuneParamAddr
        case osc2WaveformParamAddr
        case mixOsc1VolumeParamAddr
        case mixOsc2VolumeParamAddr
        case mixNoiseVolumeParamAddr
    }
    
    typealias AUParameterDescription = (String, String, ParamAddr, Float, Float, AudioUnitParameterUnit, [String]?)
    
    // MARK: Overrided properties
    override public var currentPreset: AUAudioUnitPreset? {
        get {
            if curPresetIndex < 0 {
                let userPreset = AUAudioUnitPreset()
                userPreset.number = curPresetIndex
                userPreset.name   = curPresetName
                return userPreset
            }
            else {
                return AUAudioUnitPreset(with: factoryPresetsManager.getPreset(withIndex: curPresetIndex))
            }
        }
        set {
            guard let newVal = newValue else { return }
            curPresetIndex = newVal.number
            
            if curPresetIndex < 0 {
                // Parameters will be updated using fullState
                curPresetName = newVal.name
            }
            else if let factoryPreset = factoryPresetsManager.getPreset(withIndex: curPresetIndex) {
                curPresetName = factoryPreset.presetName
                parameterTree.allParameters.forEach { param in
                    guard let val = factoryPreset.presetValue(for: param.identifier) else { return }
                    param.value = val
                }
            }
        }
    }
    
    override public var factoryPresets: [AUAudioUnitPreset]? {
        get {
            return factoryPresetsManager.allPresets().compactMap { AUAudioUnitPreset(with: $0) }
        }
    }
    
    override public var fullState: [String:Any]? {
        get {
            let preset = MinimoogInstrumentPreset(presetIndex: curPresetIndex, presetName: curPresetName, parameters: parameterTree.allParameters)
            return preset.fullState
        }
        set {
            let preset = MinimoogInstrumentPreset(with: newValue)
            parameterTree.allParameters.forEach { param in
                guard let val = preset.presetValue(for: param.identifier) else { return }
                param.value = val
            }
        }
    }
    
    override public var parameterTree: AUParameterTree {
        get { return self.curParameterTree }
        set { self.curParameterTree = newValue }
    }
    
    override public var inputBusses : AUAudioUnitBusArray {
        get { return self.curInputBusses }
        set { self.curInputBusses = newValue }
    }
    
    override public var outputBusses : AUAudioUnitBusArray {
        get { return self.curOutputBusses }
        set { self.curOutputBusses = newValue }
    }
    
    override public var internalRenderBlock : AUInternalRenderBlock {
        get { return self.minimoogInstrumentWrapper.internalRenderBlock() }
    }
    
    // MARK: Private variables
    private let paramsDescription : [AUParameterDescription] = [
        ("osc1Range"     ,"Oscillator 1 Range"       , .osc1RangeParamAddr     ,  0,  5, .indexed, ["LO","32'","16'","8'","4'","2'"]),
        ("osc1Waveform"  ,"Oscillator 1 Waveform"    , .osc1WaveformParamAddr  ,  0,  5, .indexed, ["Triangle","Ramp","Sawtooth","Square","Pulse1","Pulse2"]),
        ("osc2Range"     ,"Oscillator 2 Range"       , .osc2RangeParamAddr     ,  0,  5, .indexed, ["LO","32'","16'","8'","4'","2'"]),
        ("osc2Detune"    ,"Oscillator 2 Detune"      , .osc2DetuneParamAddr    , -8,  8, .cents  , nil),
        ("osc2Waveform"  ,"Oscillator 2 Waveform"    , .osc2WaveformParamAddr  ,  0,  5, .indexed, ["Triangle","Ramp","Sawtooth","Square","Pulse1","Pulse2"]),
        ("mixOsc1Volume" ,"Mixer Oscillator 1 Volume", .mixOsc1VolumeParamAddr ,  0, 10, .customUnit, nil),
        ("mixOsc2Volume" ,"Mixer Oscillator 2 Volume", .mixOsc2VolumeParamAddr ,  0, 10, .customUnit, nil),
        ("mixNoiseVolume","Mixer Noise Volume"       , .mixNoiseVolumeParamAddr,  0, 10, .customUnit, nil)]
    
    private var minimoogInstrumentWrapper: MinimoogInstrumentObjcWrapper
    private var curParameterTree         : AUParameterTree!
    private var curInputBusses           : AUAudioUnitBusArray!
    private var curOutputBusses          : AUAudioUnitBusArray!
    private var factoryPresetsManager    : MinimoogInstrumentFactoryPresetsManager
    
    private var curPresetIndex           : Int // Positive - factory, negative - user
    private var curPresetName            : String
    
    // MARK: Methods
    override init(componentDescription: AudioComponentDescription, options: AudioComponentInstantiationOptions = []) throws {
        // -- Set class own properties --
        self.minimoogInstrumentWrapper = MinimoogInstrumentObjcWrapper()
        self.factoryPresetsManager     = MinimoogInstrumentFactoryPresetsManager()
        self.curPresetIndex            = 0
        self.curPresetName             = ""
        
        try super.init(componentDescription:componentDescription, options:options)
        
        // -- Set inherited and overriden properties --
        self.maximumFramesToRender = 512
        
        let params = paramsDescription.map { description in
            AUParameterTree.createParameter(withIdentifier: description.0,
                                                      name: description.1,
                                                   address: description.2.rawValue,
                                                       min: AUValue(description.3),
                                                       max: AUValue(description.4),
                                                      unit: description.5,
                                                  unitName: nil,
                                                     flags: [.flag_IsWritable, .flag_IsReadable],
                                              valueStrings: description.6,
                                       dependentParameters: nil)
        }
        self.parameterTree = AUParameterTree.createTree(withChildren: params)
        
        guard let defaultFormat = AVAudioFormat(standardFormatWithSampleRate:44100.0, channels:2) else {
            throw "Invalid audio format"
        }
        
        //_audioStreamBasicDescription = *defaultFormat.streamDescription
        let inputBus      = try AUAudioUnitBus(format:defaultFormat)
        let outputBus     = try AUAudioUnitBus(format:defaultFormat)
        self.inputBusses  = AUAudioUnitBusArray(audioUnit:self, busType:.input, busses: [inputBus])
        self.outputBusses = AUAudioUnitBusArray(audioUnit:self, busType:.output, busses: [outputBus])
        
        // -- Customization and final setup --
        self.minimoogInstrumentWrapper.setSampleRate(defaultFormat.sampleRate)
        
        // A function to provide string representations of parameter values.
        self.parameterTree.implementorStringFromValueCallback = { param, valuePtr in
            let value = (valuePtr == nil ? param.value : valuePtr!.pointee)
            if (param.unit == .indexed) {
                return param.valueStrings![Int(value)]
            }
            else {
                return String(format:".2", value)
            }
        }
        
        // observe parameters change and update synth core
        self.parameterTree.implementorValueObserver = { [weak self] param, value in
            self?.minimoogInstrumentWrapper.setParameter(param.address, value:value)
        }
        
        self.parameterTree.implementorValueProvider = { [weak self] param in
            return self?.minimoogInstrumentWrapper.getParameter(param.address) ?? AUValue(0)
        }
        
        // apply default preset
        self.currentPreset = AUAudioUnitPreset(with: self.factoryPresetsManager.defaultPreset())
    }

    // MARK: AUAudioUnit Overrides
    override public func allocateRenderResources() throws {
        try super.allocateRenderResources()
        guard self.minimoogInstrumentWrapper.allocateRenderResources(musicalContext: self.musicalContextBlock ,
                                                                   outputEventBlock: self.midiOutputEventBlock,
                                                                transportStateBlock: self.transportStateBlock ,
                                                                          maxFrames: self.maximumFramesToRender)
        else { throw "Failed to allocate render resources" }
    }

    override public func deallocateRenderResources() {
        super.deallocateRenderResources()
        self.minimoogInstrumentWrapper.deallocateRenderResources()
    }
}
